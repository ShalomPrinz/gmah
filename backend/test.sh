#!/bin/bash

if [[ $1 == "--coverage" ]]; then
    coverage run -m unittest discover tests
    coverage html
else
    python -m unittest discover tests
fi
