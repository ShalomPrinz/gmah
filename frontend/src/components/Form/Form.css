:root {
  --form-input-color: hsl(218, 23%, 23%);
  --form-input-invalid-border: hsl(355, 100%, 75%);
}

.form-text-input {
  padding: 10px;
  border: 2px solid black;
  background-color: white;
  color: var(--form-input-color);
  border-radius: 10px;
}

.form-input-invalid {
  border: 4px solid var(--form-input-invalid-border);
  box-shadow: none;
}

.form-text-input:focus {
  border-color: #6c9ce0;
  box-shadow: 0 0 0 0.4rem rgba(108, 156, 224, 0.25);
  outline: none;
}
