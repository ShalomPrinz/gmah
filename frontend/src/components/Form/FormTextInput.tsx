import { useField } from "formik";

export type TextInput = {
  name: string;
};

interface FormTextInputProps {
  name: string;
}

export const FormTextInput = ({ name }: FormTextInputProps) => {
  const [field, meta] = useField(name);
  return (
    <div className="d-flex flex-column">
      <label className="fs-5" htmlFor={name}>
        {name}
      </label>
      <input
        className="form-text-input p-2 m-3 mb-2"
        type="text"
        {...field}
      />
      {meta.touched && meta.error ? (
        <div className="text-danger">{meta.error}</div>
      ) : null}
    </div>
  );
};
