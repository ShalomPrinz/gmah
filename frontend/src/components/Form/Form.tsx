import { Form, Formik } from "formik";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";

import { ConditionalList } from "../";
import type { TextInput } from "./FormTextInput";
import { FormTextInput } from "./FormTextInput";
import "./Form.css";

interface FormProps {
  onSubmit: (values: {}) => Promise<any>;
  /** Yup Object Schema */
  schema: any;
  textInputs: Array<TextInput>;
  title: string;
}

const inputCallback = (item: TextInput) => (
  <Col>
    <FormTextInput {...item} />
  </Col>
);

const AppForm = ({ onSubmit, schema, textInputs, title }: FormProps) => {
  const initialValues = textInputs.reduce(
    (o, key) => ({ ...o, [key.name]: "" }),
    {}
  );

  // @ts-ignore This function parameters types is determined by Formik
  const handleSubmit = (values, { resetForm, setSubmitting }) =>
    onSubmit(values)
      .then((submitSucceed) => {
        if (submitSucceed) resetForm();
      })
      .finally(() => setSubmitting(false));

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={schema}
      onSubmit={handleSubmit}
    >
      <Form className="container">
        <Row>
          <h1>{title}</h1>
        </Row>
        <Row>
          <ConditionalList itemCallback={inputCallback} list={textInputs} />
        </Row>
        <Row>
          <button
            className="fs-1 bg-default w-50 mx-auto p-3 m-4 rounded button-border-focus"
            type="submit"
          >
            {title}
          </button>
        </Row>
      </Form>
    </Formik>
  );
};

export default AppForm;
