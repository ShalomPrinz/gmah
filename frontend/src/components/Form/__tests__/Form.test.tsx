import { describe, expect, it, vi } from "vitest";
import { number, object, string } from "yup";

import { render, screen, setupUser } from "../../../../test";

import Form from "../Form";

const emptyOnSubmit = () => Promise.resolve();

describe("Form", () => {
  describe("snapshot", () => {
    it("should render empty form", () => {
      const { asFragment } = render(
        <Form
          onSubmit={emptyOnSubmit}
          schema={object()}
          textInputs={[]}
          title=""
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render form with one text input", () => {
      const { asFragment } = render(
        <Form
          onSubmit={emptyOnSubmit}
          schema={object({
            name: string(),
          })}
          textInputs={[
            {
              label: "Name",
              path: "name",
            },
          ]}
          title="Basic Form"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render form with three text inputs", () => {
      const { asFragment } = render(
        <Form
          onSubmit={emptyOnSubmit}
          schema={object({
            name: string(),
            value: number(),
            content: string(),
          })}
          textInputs={[
            {
              label: "Name",
              path: "name",
            },
            {
              label: "Value",
              path: "value",
            },
            {
              label: "Content",
              path: "content",
            },
          ]}
          title="Three Inputs Form"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should ignore double size input if inputsInRow not set to '6'", () => {
      const { asFragment } = render(
        <Form
          onSubmit={emptyOnSubmit}
          schema={object()}
          textInputs={[
            {
              label: "Name",
              path: "name",
              doubleSize: true,
            },
          ]}
          title=""
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render double size input if inputsInRow set to '6'", () => {
      const { asFragment } = render(
        <Form
          onSubmit={emptyOnSubmit}
          schema={object()}
          textInputs={[
            {
              label: "Name",
              path: "name",
              doubleSize: true,
            },
          ]}
          title=""
          inputsInRow="6"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render form with unique submit text", () => {
      const { asFragment } = render(
        <Form
          onSubmit={emptyOnSubmit}
          schema={object({
            name: string(),
          })}
          textInputs={[
            {
              label: "Name",
              path: "name",
            },
          ]}
          title="Basic Form"
          submitText="Submit This Form"
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("behavior", () => {
    it("should submit initial data", async () => {
      const user = setupUser();
      const onSubmit = vi.fn((values: object) => Promise.resolve());
      const initialData = { id: 0, value: "Some data" };
      render(
        <Form
          onSubmit={onSubmit}
          schema={object()}
          textInputs={[]}
          title="Initial Data Form"
          initialData={initialData}
        />
      );

      const submitButton = screen.getByRole("button");
      await user.click(submitButton);

      expect(onSubmit).toBeCalledWith(initialData);
    });

    it("should submit typed data", async () => {
      const user = setupUser();
      const onSubmit = vi.fn((values: object) => Promise.resolve());
      render(
        <Form
          onSubmit={onSubmit}
          schema={object()}
          textInputs={[
            {
              label: "Full Name",
              path: "name",
            },
          ]}
          title="Initial Data Form"
        />
      );

      const dataToType = "Shalom Prinz";
      const inputElement = screen.getByRole("textbox");
      await user.type(inputElement, dataToType);

      const submitButton = screen.getByRole("button");
      await user.click(submitButton);

      expect(onSubmit).toBeCalledWith({ name: dataToType });
    });

    it("should submit typed data over initial data", async () => {
      const user = setupUser();
      const onSubmit = vi.fn((values: object) => Promise.resolve());
      render(
        <Form
          onSubmit={onSubmit}
          schema={object()}
          textInputs={[
            {
              label: "Full Name",
              path: "name",
            },
          ]}
          title="Initial Data Form"
          initialData={{ name: "some placeholder" }}
        />
      );

      const dataToType = "Real Name";
      const inputElement = screen.getByRole("textbox");
      await user.clear(inputElement);
      await user.type(inputElement, dataToType);

      const submitButton = screen.getByRole("button");
      await user.click(submitButton);

      expect(onSubmit).toBeCalledWith({ name: dataToType });
    });

    it("should not submit if schema validation failed", async () => {
      const user = setupUser();
      const onSubmit = vi.fn((values: object) => Promise.resolve());
      const expectedString = "Shalom Prinz";
      render(
        <Form
          onSubmit={onSubmit}
          schema={object({
            name: string().matches(new RegExp(`^${expectedString}$`)),
          })}
          textInputs={[
            {
              label: "Full Name",
              path: "name",
            },
          ]}
          title=""
        />
      );

      const schemaFailureText = "Some Text";
      const inputElement = screen.getByRole("textbox");
      await user.type(inputElement, schemaFailureText);

      const submitButton = screen.getByRole("button");
      await user.click(submitButton);

      expect(onSubmit).not.toBeCalled();

      await user.clear(inputElement);
      await user.type(inputElement, expectedString);
      await user.click(submitButton);

      expect(onSubmit).toBeCalledWith({ name: expectedString });
    });

    it("should display error in the DOM if schema validation failed", async () => {
      const user = setupUser();
      const onSubmit = vi.fn((values: object) => Promise.resolve());
      const { asFragment } = render(
        <Form
          onSubmit={onSubmit}
          schema={object({
            name: string().matches(/"password"/),
          })}
          textInputs={[
            {
              label: "Full Name",
              path: "name",
            },
          ]}
          title=""
        />
      );

      const inputElement = screen.getByRole("textbox");
      await user.type(inputElement, "wrong");
      await user.tab(); // Unfocus element should cause validation

      expect(asFragment()).toMatchSnapshot();

      await user.clear(inputElement);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
