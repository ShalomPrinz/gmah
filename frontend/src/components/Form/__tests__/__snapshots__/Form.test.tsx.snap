// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form > behavior > should display error in the DOM if schema validation failed 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1 />
    </div>
    <div
      class="d-flex justify-content-center row"
    >
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="name"
          >
            Full Name
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center form-input-invalid"
            name="name"
            title="name must match the following: \\"/\\"password\\"/\\""
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form > behavior > should display error in the DOM if schema validation failed 2`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1 />
    </div>
    <div
      class="d-flex justify-content-center row"
    >
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="name"
          >
            Full Name
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="name"
            title="Full Name"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form > snapshot > should ignore double size input if inputsInRow not set to '6' 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1 />
    </div>
    <div
      class="d-flex justify-content-center row"
    >
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="name"
          >
            Name
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="name"
            title="Name"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form > snapshot > should render double size input if inputsInRow set to '6' 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1 />
    </div>
    <div
      class="d-flex justify-content-center row"
    >
      <div
        class="col-md-4"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="name"
          >
            Name
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="name"
            title="Name"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form > snapshot > should render empty form 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1 />
    </div>
    <div
      class="d-flex justify-content-center row"
    />
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form > snapshot > should render form with one text input 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1>
        Basic Form
      </h1>
    </div>
    <div
      class="d-flex justify-content-center row"
    >
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="name"
          >
            Name
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="name"
            title="Name"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      >
        Basic Form
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form > snapshot > should render form with three text inputs 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1>
        Three Inputs Form
      </h1>
    </div>
    <div
      class="d-flex justify-content-center row"
    >
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="name"
          >
            Name
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="name"
            title="Name"
            type="text"
          />
        </div>
      </div>
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="value"
          >
            Value
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="value"
            title="Value"
            type="text"
          />
        </div>
      </div>
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="content"
          >
            Content
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="content"
            title="Content"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      >
        Three Inputs Form
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Form > snapshot > should render form with unique submit text 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <h1>
        Basic Form
      </h1>
    </div>
    <div
      class="d-flex justify-content-center row"
    >
      <div
        class="col-md-3"
      >
        <div
          class="d-flex flex-column my-3"
        >
          <label
            class="fs-4 my-2"
            for="name"
          >
            Name
          </label>
          <input
            class="fs-5 form-text-input p-1 m-1 text-center"
            name="name"
            title="Name"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <button
        class="fs-1 bg-default w-50 mx-auto p-3 mt-5 rounded button-border-focus"
        type="submit"
      >
        Submit This Form
      </button>
    </div>
  </form>
</DocumentFragment>
`;
